TUIButton Revamped!

avaidyam opened the pull request:
The current `TUIButton` has several flaws, aside from not even being "API-complete":
- The button style does not have OS X version consistency: 10.6, 10.7, and 10.8+ all have different button drawing.
- The button style does not have User-wide consistency: the user will notice quirky buttons easily.
- The button style does not have proper HiDPI capable drawing (`NSShadow` can go die in a ditch :cry: ).

After some investigation, I've determined that AppKit's demon-child `NSCell` is actually quite useful for us: We can use these little drawing blocks on top of our amazing `CALayer` backed `TUIView`s easily and keep OS and User consistency.

`NSButtonCell` uses the `CoreUI.framework` introduced in 10.5 (and put to use in 10.7) to draw all its controls perfectly, and can do everything for us: so why do it ourselves? Now `TUIButton` has a shared button renderer (an `NSButtonCell`) whose bezel style is swapped out and is redrawn for each button.

Now the `TUIButton` is indiscernible from a standard `NSButton`! I've yet to figure out how to pulse the button though: my idea is to use a `CAAnimation` to swap button colors, but there's no way to find the "darker" button color in the AppKit pulse. AppKit also uses a special `NSUIHeartBeat` thread for these animations.

EDIT: It now also replaces `NSPopUpButtonCell` by extending the menu functionality it previously had! One button to rule them all!

avaidyam committed:
Clean up TUIButton.

avaidyam committed:
Add TUIButtonTypeStandard support.

Mimics the Calendar.app popover buttons. Simple.

avaidyam committed:
Added TUIButtonTypeBeveled.

And a few additional drawing facilities for the rest of the button
types.

avaidyam committed:
Added TUIButtonTypeEmbossed and TUIButtonTypeGradient.

avaidyam committed:
Added TUIButtonTypeMinimal and TUIButtonTypeInline.

avaidyam committed:
A few fixes and changes.

avaidyam committed:
Added better examples of the buttons.

avaidyam committed:
Fixed many odd bugs.

avaidyam committed:
Merge remote-tracking branch 'upstream/master' into upstream-button-additions

Conflicts:
	TwUI.xcodeproj/project.pbxproj

avaidyam committed:
Merge remote-tracking branch 'upstream/master' into upstream-button-additions

Conflicts:
	ExampleProject/Example.xcodeproj/project.pbxproj
	TwUI.xcodeproj/project.pbxproj

avaidyam committed:
Button cleanup.

avaidyam committed:
Fixed control state changes.

avaidyam committed:
Cleaned up TUIControl.

avaidyam committed:
TUIButton and TUIControl documentation.

avaidyam committed:
Add TUIButtonTypeClear.

avaidyam committed:
quick!

avaidyam committed:
Merge remote-tracking branch 'upstream/master' into upstream-button-additions

Conflicts:
	TwUI.xcodeproj/project.pbxproj
	lib/UIKit/TUIButton.h
	lib/UIKit/TUIControl.h
	lib/UIKit/TUIKit.h

avaidyam committed:
Better example.

avaidyam committed:
Clean up Example project.

avaidyam committed:
Fix example project.

avaidyam committed:
Switch to CoreUI-automated drawing.

This stays consistent with system controls and OS version as well.

avaidyam committed:
Highly efficient shared button renderer.

avaidyam committed:
Added concise and thorough button documentation.

jwilling commented on ExampleProject/Example/ExampleAppDelegate.h:
```diff
@@ -17,13 +17,12 @@
-{
-  NSWindow    * tableViewWindow;
-  NSWindow    * scrollViewWindow;
+@interface ExampleAppDelegate : NSObject <NSApplicationDelegate> {
```
All of the demo changes not pertaining to _buttons_ **only** should not be in this pull request.

avaidyam replied:
All I did was clean up the example code though... nothing new introduced.

jwilling replied:
Clean up can still introduce bugs, and this pull request is far too large in its current state to merge.

jwilling commented on lib/UIKit/NSShadow+TUIExtensions.m:
```diff
@@ -19,9 +19,13 @@
 @implementation NSShadow (TUIExtensions)
 
 + (NSShadow *)tui_shadowWithRadius:(CGFloat)radius offset:(CGSize)offset color:(NSColor *)color {
+	CGFloat scale = 1.0f;
```
The shadow changes should be in a separate pull request.

jwilling replied:
This needs to be in a PR before this can be opened again.

avaidyam replied:
GitHub isn't too helpful here: you can only see branch changes at my branch or by reopening this PR. galaxas0/twui/upstream-button-additions

jwilling commented on lib/UIKit/TUICGAdditions.h:
```diff
@@ -71,3 +71,5 @@
  */
 extern NSData *TUIGraphicsDrawAsPDF(CGRect *optionalMediaBox, void(^draw)(CGContextRef));
+
+extern CGImageRef TUIGenerateNoiseImage(CGSize size, CGFloat factor) CF_RETURNS_RETAINED;
```
This addition should be in a different pull request, also.

avaidyam replied:
Whoah, that's odd- I thought I stashed that change. I'll fix that.

jwilling replied:
This still needs to be in a PR.

jwilling commented on lib/UIKit/TUIControl.h:
```diff
@@ -17,67 +17,159 @@
 
 typedef enum TUIControlEvents : NSUInteger {
+	
+	// A mouse down event in the control.
```
These `TUIControl` changes seem to be independent of the button changes, and should therefore be in a separate PR.

avaidyam replied:
Actually, I just documented and exposed a few `TUIControl` things to make the `TUIButton` function the way it does. No new `TUIControl` features (aside from system control tint notifications… which are for the buttons too.)

jwilling replied:
I'm sorry, but I see some features in here that are definitely not just documentation. This can exist independently of `TUIButton`, so it should be split.

avaidyam committed:
Reverted unneccessary changes.

jwilling commented:
As I've mentioned, this PR is far too large to merge. I appreciate the work you've done here, but  until it is split up significantly into logical parts this will remain closed.

avaidyam committed:
Cleaned up the Example project.

avaidyam committed:
Added relative image-text positioning (to all controls, but implemented in TUIButton).

avaidyam committed:
Button fixes.

avaidyam committed:
Image drawing updates.

avaidyam committed:
Improved drawing refactoring.

avaidyam commented:
@jwilling Yep, the commits are there alright!

jwilling commented:
Please merge from master. This might perhaps need a rebase.

avaidyam committed:
Merge remote-tracking branch 'upstream/master' into upstream-button-additions

Conflicts:
	lib/UIKit/TUIButton+Content.m
	lib/UIKit/TUIButton.m
	lib/UIKit/TUIControl.h
	lib/UIKit/TUIControl.m

avaidyam commented:
@jwilling I've merged away! :]

jwilling commented on lib/UIKit/TUITextView.m:
```diff
@@ -811,7 +811,7 @@ TUIViewDrawRect TUITextViewStandardFrame(void)
 		NSRect blackOutlineFrame = NSMakeRect(0.0, 1.0, bounds.size.width, bounds.size.height-2.0);
 		NSGradient *gradient = nil;
-		if([NSApp isActive]) {
+		if(![NSApp isActive]) {
```
Even though this is a super small change, it doesn't pertain to the pull request. Can you pop this one in a separate PR?

avaidyam replied:
That's another one that got in there somehow... I'll do that PR later. I've got simply way too many branches needing work.

jwilling commented:
I will attempt to give a review in the near future, but I have limited time. Hopefully soon I can get to this.

avaidyam commented:
That's alright, there's no compulsion to be done with it right this minute. :]

avaidyam committed:
Clean up fixes.

avaidyam committed:
One button to rule them all!

Extended to support popups and pulldowns. (replaces NSPopupButtonCell
and NSButtonCell both at once!)

avaidyam commented:
@jwilling When you're reviewing this PR again, please do check out the menu functionality! I want to make sure every single quirk is fixed in this button update.

avaidyam committed:
Remove button from XIB

avaidyam committed:
Extend button examples.

avaidyam committed:
Fix stuff.

avaidyam committed:
Refactoring and button fixes.

avaidyam committed:
Finalize button menu system.

avaidyam committed:
Reverted button tests.

jspahrsummers commented:
> After some investigation, I've determined that AppKit's demon-child `NSCell` is actually quite useful for us: We can use these little drawing blocks on top of our amazing `CALayer` backed `TUIView`s easily and keep OS and User consistency.

`NSCell` does more than just drawing. I'm skeptical that ignoring its event handling routines will actually work correctly in all cases.

> It now also replaces `NSPopUpButtonCell` by extending the menu functionality it previously had! One button to rule them all!

Out of scope. I'm a bit confused about why this was added _after_ @jwilling pointed out that the pull request was already too large.

avaidyam commented:
@jspahrsummers I believe the reason @jwilling pointed that out was because of the non-`TUIButton` class changes (Retina shadows, `TUIControl` updates, etc.)

avaidyam commented:
@jspahrsummers If you believe that this PR is still too large (which I'm sure it probably is), I don't mind it being closed. :construction:

jspahrsummers commented:
Changes to the same class can still comprise different logical parts – in this case, refactoring the basic functionality of the class should be one pull request, and adding menu support would be another, later one (though, admittedly, it's unlikely to make it into github/twui).

avaidyam commented:
@jspahrsummers Do you suggest I break this PR down into multiple ones? The reason I haven't done that was because they would be very interdependent, and no one likes cyclic dependency PRs. :(

avaidyam committed:
Fix button example.

jspahrsummers commented:
> If you believe that this PR is still too large (which I'm sure it probably is), I don't mind it being closed. :construction: 

I'm not opposed to this PR on principle, but the use of `NSCell` is worrisome, and the pop-up menu stuff definitely should be moved out. I'm afraid I don't have time to review it in detail, though, and I'll leave to others the decision to close or review it.

> Do you suggest I break this PR down into multiple ones? The reason I haven't done that was because they would be very interdependent, and no one likes cyclic dependency PRs. :(

Why would they be interdependent? If the menu stuff was added in a few commits _after_ most of the work was done, the dependency seems pretty unidirectional.

So, the menu stuff is the major logical splitting point I see here, especially given that it was tacked on to a PR already in review. Unfortunately, all of the repo collaborators are pretty short on TwUI time these days, so the menu additions shouldn't be PRed at this point.

avaidyam commented:
@jspahrsummers I can cherry-pick the menu stuff out into a separate PR then. For now, I think it's better to leave this PR as is.

jspahrsummers commented:
> I can cherry-pick the menu stuff out into a separate PR then.

You're welcome to cherry-pick it into another branch on your fork, but please don't open a new pull request with it (again, due to limited reviewer time and maintenance mode).

avaidyam commented:
@jspahrsummers Understood! Thanks :smiley:

jspahrsummers commented:
Upon a cursory review, this seems pretty dependent on the `NSCell` functionality. Unless you can remove the use of `NSCell` entirely, this seems too risky to merge.

avaidyam commented:
@jspahrsummers That would be far too difficult... I think it would be better to simply close the PR then. This would require far more maintenance on github's part.

jspahrsummers commented:
Okay. Thank you for understanding. :star2:
